File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/howto/samples/xml/readxmlstream/cs/ReadXmlStream.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft .NET SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//-----------------------------------------------------------------------

namespace HowTo.Samples.XML
{

using System;
using System.IO;
using System.Xml;

public class ReadXmlStreamSample
{
    public static void Main()
    {
        ReadXmlStreamSample myReadXmlStreamSample = new ReadXmlStreamSample();
        myReadXmlStreamSample.Run();
    }

    public void Run()
    {
        StringReader stream;
        XmlTextReader reader = null;

        try
        {
            Console.WriteLine ("Initializing StringReader ...");

            stream = new StringReader("<?xml version='1.0'?>" +
                                      "<!-- This file represents a fragment of a book store inventory database -->" +
                                      "<bookstore>" +
                                      " <book genre=\"autobiography\" publicationdate=\"1981\" ISBN=\"1-861003-11-0\">" +
                                      "   <title>The Autobiography of Benjamin Franklin</title>" +
                                      "   <author>" +
                                      "       <first-name>Benjamin</first-name>" +
                                      "       <last-name>Franklin</last-name>" +
                                      "   </author>" +
                                      "   <price>8.99</price>" +
                                      " </book>" +
                                      " <book genre=\"novel\" publicationdate=\"1967\" ISBN=\"0-201-63361-2\">" +
                                      "   <title>The Confidence Man</title>" +
                                      "   <author>" +
                                      "       <first-name>Herman</first-name>" +
                                      "       <last-name>Melville</last-name>" +
                                      "   </author>" +
                                      "   <price>11.99</price>" +
                                      " </book>" +
                                      "  <book genre=\"philosophy\" publicationdate=\"1991\" ISBN=\"1-861001-57-6\">" +
                                      "   <title>The Gorgias</title>" +
                                      "   <author>" +
                                      "       <name>Plato</name>" +
                                      "   </author>" +
                                      "   <price>9.99</price>" +
                                      " </book>" +
                                      "</bookstore>");

            // Load the XmlTextReader from the stream
            reader = new XmlTextReader (stream);

            Console.WriteLine ("Processing ...");
            Console.WriteLine ();
            FormatXml(reader);
        }

        catch (Exception e)
        {
            Console.WriteLine ("Exception: {0}", e.ToString());
        }

        finally
        {
            Console.WriteLine();
            Console.WriteLine("Processing of stream complete.");
            // Finished with XmlTextReader
            if (reader != null)
                reader.Close();
        }
    }

private static void FormatXml (XmlReader reader)
{
    int piCount=0, docCount=0, commentCount=0, elementCount=0, attributeCount=0, textCount=0, whitespaceCount=0;

    while (reader.Read())
    {
            switch (reader.NodeType)
            {
            case XmlNodeType.ProcessingInstruction:
                Format (reader, "ProcessingInstruction");
                piCount++;
                break;
            case XmlNodeType.DocumentType:
                Format (reader, "DocumentType");
                docCount++;
                break;
            case XmlNodeType.Comment:
                Format (reader, "Comment");
                commentCount++;
                break;
            case XmlNodeType.Element:
                Format (reader, "Element");
                while(reader.MoveToNextAttribute())
                {
                    Format (reader, "Attribute");
                }
                elementCount++;
                if (reader.HasAttributes)
                    attributeCount += reader.AttributeCount;
                break;
            case XmlNodeType.Text:
                Format (reader, "Text");
                textCount++;
                break;
            case XmlNodeType.Whitespace:
                whitespaceCount++;
                break;
            }
        }

        // Display the Statistics
        Console.WriteLine ();
        Console.WriteLine("Statistics for stream");
        Console.WriteLine ();
        Console.WriteLine("ProcessingInstruction: {0}",piCount++);
        Console.WriteLine("DocumentType: {0}",docCount++);
        Console.WriteLine("Comment: {0}",commentCount++);
        Console.WriteLine("Element: {0}",elementCount++);
        Console.WriteLine("Attribute: {0}",attributeCount++);
        Console.WriteLine("Text: {0}",textCount++);
        Console.WriteLine("Whitespace: {0}",whitespaceCount++);
    }

    // Format the output
    private static void Format(XmlReader reader, String nodeType)
    {
        // Format the output
        Console.Write(reader.Depth + " ");
        Console.Write(reader.AttributeCount + " ");
        for (int i=0; i < reader.Depth; i++)
        {
            Console.Write('\t');
        }

        Console.Write(nodeType + "<" + reader.Name + ">" + reader.Value);
        Console.WriteLine();
    }

} // End class ReadXmlStreamSample
} // End namespace HowTo.Samples.XML

File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/howto/samples/xml/readxmlstream/vb/ReadXmlStream.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft .NET SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'-----------------------------------------------------------------------

Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Xml

namespace HowTo.Samples.XML

public class ReadXmlStreamSample

    Shared sub Main
        Dim myReadXmlStreamSample as ReadXmlStreamSample
        myReadXmlStreamSample = new ReadXmlStreamSample()
        myReadXmlStreamSample.Run()
    end sub

    public sub Run()

        Dim stream as StringReader
        Dim reader as XmlTextReader = nothing

        try
            Console.WriteLine ("Initializing StringReader ...")
            stream = new StringReader ("<?xml version='1.0'?>" & _
                                       "<!-- This file represents a fragment of a book store inventory database -->" & _
                                       "<bookstore>" & _
                                       " <book genre=""autobiography"" publicationdate=""1981"" ISBN=""1-861003-11-0"">" & _
                                       "   <title>The Autobiography of Benjamin Franklin</title>" & _
                                       "   <author>" & _
                                       "       <first-name>Benjamin</first-name>" & _
                                       "       <last-name>Franklin</last-name>" & _
                                       "   </author>" & _
                                       "   <price>8.99</price>" & _
                                       " </book>" & _
                                       " <book genre=""novel"" publicationdate=""1967"" ISBN=""0-201-63361-2"">" & _
                                       "   <title>The Confidence Man</title>" & _
                                       "   <author>" & _
                                       "       <first-name>Herman</first-name>" & _
                                       "       <last-name>Melville</last-name>" & _
                                       "   </author>" & _
                                       "   <price>11.99</price>" & _
                                       " </book>" & _
                                       "  <book genre=""philosophy"" publicationdate=""1991"" ISBN=""1-861001-57-6"">" & _
                                       "   <title>The Gorgias</title>" & _
                                       "   <author>" & _
                                       "       <name>Plato</name>" & _
                                       "   </author>" & _
                                       "   <price>9.99</price>" & _
                                       " </book>" & _
                                       "</bookstore>")

            ' Load the XmlTextReader from the stream
            reader = new XmlTextReader (stream)

            Console.WriteLine ("Processing ...")
            Console.WriteLine ()
            FormatXml(reader)

        catch e as Exception
            Console.WriteLine ("Exception: {0} ", e.ToString())

        finally
            Console.WriteLine()
            Console.WriteLine("Processing of stream complete.")
            If Not reader Is Nothing
                reader.Close()
            end if

        end try
    end sub

    private shared sub FormatXml (reader as XmlTextReader)

        Dim piCount, docCount, commentCount, elementCount as Integer
        Dim attributeCount, textCount, whitespaceCount as Integer

        While reader.Read()

            Select (reader.NodeType)

            case XmlNodeType.ProcessingInstruction:
                Format (reader, "ProcessingInstruction")
                piCount += 1

            case XmlNodeType.DocumentType:
                Format (reader, "DocumentType")
                docCount += 1

            case XmlNodeType.Comment:
                Format (reader, "Comment")
                commentCount += 1

            case XmlNodeType.Element:
                Format (reader, "Element")
                elementCount += 1
                While reader.MoveToNextAttribute()
                    Format (reader, "Attribute")
                end While

                if (reader.HasAttributes)
                    attributeCount += reader.AttributeCount
                end if

            case XmlNodeType.Text:
                Format (reader, "Text")
                textCount += 1

            case XmlNodeType.Whitespace:
                whitespaceCount += 1

            end Select

        end While

        ' Display the Statistics for the file
        Console.WriteLine ()
        Console.WriteLine("Statistics for stream")
        Console.WriteLine ()
        Console.WriteLine("ProcessingInstruction: " & piCount)
        Console.WriteLine("DocumentType: " & docCount)
        Console.WriteLine("Comment: " & commentCount)
        Console.WriteLine("Element: " & elementCount)
        Console.WriteLine("Attribute: " & attributeCount)
        Console.WriteLine("Text: " & textCount)
        Console.WriteLine("Whitespace: " & whitespaceCount)
    end sub

    private shared sub Format(byref reader as XmlTextReader, nodeType as String)

        ' Format the output
        Console.Write(reader.Depth & " ")
        Console.Write(reader.AttributeCount & " ")

        Dim i as Integer
        for i = 0 to reader.Depth - 1
            Console.Write(Strings.chr(9))
        Next

        Console.Write(reader.Prefix & nodeType & "<" & reader.Name & ">" & reader.Value)
        Console.WriteLine()

    end sub

end Class
end Namespace
